File: lectures/lecture-02-what-a-harness-actually-is/exercises/exercise-02-ablation-report/starter/typescript/main.ts
// ablation-report exercise, TypeScript starter.
//
// The aggregation plumbing works; the comparison logic does not. Your task:
// in compare(), fill in outcome_changed and signature per SPEC.md. Run
// ../../verify.sh --stack=typescript until it exits 0.

import { existsSync, readFileSync, statSync } from "node:fs";
import { join } from "node:path";
import { pathToFileURL } from "node:url";

const SUBSYSTEMS = ["instructions", "state", "environment", "tools", "feedback"] as const;

interface LoopReport {
  readonly disabled: string | null;
  readonly outcome: string;
  readonly issues: readonly string[];
}

function loadReport(reportsDir: string, name: string): LoopReport {
  return JSON.parse(readFileSync(join(reportsDir, name), "utf8")) as LoopReport;
}

export function compare(_baseline: LoopReport, ablated: LoopReport): Record<string, unknown> {
  // Exercise: outcome_changed is whether the ablated outcome differs from
  // the baseline outcome; signature is the ablated run's first issue
  // string, or null when it has none.
  return {
    disabled: ablated.disabled,
    outcome: ablated.outcome,
    outcome_changed: false,
    issues: ablated.issues.length,
    signature: null,
  };
}

export function buildReport(reportsDir: string): Record<string, unknown> {
  const baseline = loadReport(reportsDir, "full.json");
  const ablations = SUBSYSTEMS.map((name) =>
    compare(baseline, loadReport(reportsDir, `disable-${name}.json`)),
  );
  return {
    baseline: { outcome: baseline.outcome, issues: baseline.issues.length },
    ablations,
    all_degraded: ablations.every((entry) => entry.outcome_changed === true),
  };
}

function main(argv: readonly string[]): number {
  const reportsDir = argv[2];
  if (!reportsDir || argv.length !== 3) {
    console.error("usage: main.ts <reports-dir>");
    return 2;
  }
  if (!existsSync(reportsDir) || !statSync(reportsDir).isDirectory()) {
    console.error(`error: not a directory: ${reportsDir}`);
    return 2;
  }
  process.stdout.write(JSON.stringify(buildReport(reportsDir), null, 2) + "\n");
  return 0;
}

if (import.meta.url === pathToFileURL(process.argv[1] ?? "").href) {
  process.exit(main(process.argv));
}

File: lectures/lecture-02-what-a-harness-actually-is/exercises/exercise-02-ablation-report/starter/python/main.py
"""ablation-report exercise, Python starter.

The aggregation plumbing works; the comparison logic does not. Your task:
in compare(), fill in outcome_changed and signature per SPEC.md. Run
../../verify.sh --stack=python until it exits 0.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

SUBSYSTEMS = ("instructions", "state", "environment", "tools", "feedback")


def load_report(reports_dir: Path, name: str) -> dict:
    return json.loads((reports_dir / name).read_text(encoding="utf-8"))


def compare(baseline: dict, ablated: dict) -> dict:
    issues = ablated["issues"]
    # Exercise: outcome_changed is whether the ablated outcome differs from
    # the baseline outcome; signature is the ablated run's first issue
    # string, or None when it has none.
    return {
        "disabled": ablated["disabled"],
        "outcome": ablated["outcome"],
        "outcome_changed": False,
        "issues": len(issues),
        "signature": None,
    }


def build_report(reports_dir: Path) -> dict:
    baseline = load_report(reports_dir, "full.json")
    ablations = [
        compare(baseline, load_report(reports_dir, f"disable-{name}.json"))
        for name in SUBSYSTEMS
    ]
    return {
        "baseline": {"outcome": baseline["outcome"], "issues": len(baseline["issues"])},
        "ablations": ablations,
        "all_degraded": all(entry["outcome_changed"] for entry in ablations),
    }


def main(argv: list[str]) -> int:
    if len(argv) != 2:
        print("usage: main.py <reports-dir>", file=sys.stderr)
        return 2
    reports_dir = Path(argv[1])
    if not reports_dir.is_dir():
        print(f"error: not a directory: {reports_dir}", file=sys.stderr)
        return 2
    print(json.dumps(build_report(reports_dir), indent=2))
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))
